File: Inorder Successor in BST[30-July-2023]/Optimized Naive Approach/Solution.java
//  >> JAVA CODE: APPROACH 1 <<

class Solution
{
    // returns the inorder successor of the Node x in BST (rooted at 'root')
    public void inorder(Node root,Node x,boolean[] found,Node[] ans)
    {
        if(root==null) return;
        
        inorder(root.left,x,found,ans);
        
        if(found[0]==true)
        {
            ans[0]=root;
            found[0]=false;
        }
        if(root.data==x.data)
        {
            found[0]=true;
        }
        
        
        inorder(root.right,x,found,ans);
    }
	public Node inorderSuccessor(Node root,Node x)
    {
       boolean[] found={false};
       Node[] ans={null};
       
       inorder(root,x,found,ans);
       
       
       return ans[0];
    }
}

File: Inorder Successor in BST[30-July-2023]/Optimized Naive Approach/Solution.cpp
// >> CPP CODE: APPROACH 1 <<

class Solution{
  public:
    // returns the inorder successor of the Node x in BST (rooted at 'root')
    void inorder(Node *root,Node *x,Node** ans,bool* found)
    {
        if(root==NULL) return;
        
        inorder(root->left,x,ans,found);
        
        if(*found==true)
        {
            *ans=root;
            *found=false;
        }
        else if(root->data==x->data)
        {
            *found=true;
        }
        
        
        inorder(root->right,x,ans,found);
    }
    Node * inOrderSuccessor(Node *root, Node *x)
    {
        Node *ans=NULL;
        bool found=false;
        
        inorder(root,x,&ans,&found);
        
        return ans;
    }
};